pkg/agent/openflow/packetin.go

```diff
@@ -0,0 +1,244 @@
+	10:  "SpoofGuard",
+	20:  "ARPResponder",
+	30:  "ConntrackZone",
+	31:  "ContrackState",
+	40:  "DNAT",
+	50:  "EgressRule",
+	60:  "EgressDefaultRule",
+	70:  "L3Forwarding",
+	80:  "L2Forwarding",
+	90:  "IngressRule",
+	100: "IngressDefaultRule",
+	105: "ConntrackCommit",
+	110: "Output",
+}
+
+func (c *client) ReceivePacketIn(traceflowController interface{}, stopCh <-chan struct{}) {
+	tfController, ok := traceflowController.(TraceflowController)
+	if !ok {
+		klog.Errorf("Invalid Traceflow controller.")
+		<-stopCh
+		return
+	}
+	ch := make(chan *ofctrl.PacketIn)
+	c.SubscribePacketIn(uint8(ofprAction), ch)
+
+	wait.PollUntil(time.Second, func() (done bool, err error) {
+		pktIn := <-ch
+		oldTf, nodeResult, err := c.parsePacketIn(pktIn, tfController)
+		if err != nil {
+			klog.Errorf("parsePacketIn error: %+v", err)
+			return false, err
+		}
+		// Retry when update CRD conflict which caused by multiple agents updating one CRD at same time
+		err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			tf, err := tfController.TraceflowInformer().Lister().Get(oldTf.Name)
+			if err != nil {
+				klog.Warningf("Get traceflow failed: %+v", err)
+				return err
+			}
+			if oldTf.Status.Phase == opsv1alpha1.Succeeded {
+				tf.Status.Phase = oldTf.Status.Phase
+			}
+			tf.Status.Results = append(tf.Status.Results, *nodeResult)
+			tf, err = tfController.TraceflowClient().OpsV1alpha1().Traceflows().Update(tf)
+			if err != nil {
+				klog.Warningf("Update traceflow failed: %+v", err)
+				return err
+			}
+			tfController.UpdateTagCache(tf)
+			klog.Infof("Updated traceflow %s: %+v", tf.Name, nodeResult)
+			return err
+		})
+		if err != nil {
+			klog.Errorf("Update traceflow error: %+v", err)
+			return false, err
+		}
+		return false, err
+	}, stopCh)
+}
+
+func (c *client) parsePacketIn(pktIn *ofctrl.PacketIn, tfController TraceflowController) (*opsv1alpha1.Traceflow, *opsv1alpha1.NodeResult, error) {
+	matchers := pktIn.GetMatches()
+	var match *ofctrl.MatchField
+
+	// get data plane tag
+	if match = getMatchRegField(matchers, uint32(TraceflowReg)); match == nil {
+		return nil, nil, errors.New("traceflow data plane tag not found")
+	}
+	rngTag := openflow13.NewNXRange(int(OfTraceflowMarkRange[0]), int(OfTraceflowMarkRange[1]))
+	tag, err := getInfoInReg(match, rngTag)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// get traceflow CRD from cache by data plane tag
+	tf, err := tfController.GetRunningTraceflowCRD(uint8(tag))
+	if err != nil {
+		return nil, nil, err
+	}
+
+	obs := make([]opsv1alpha1.Observation, 0)
+	isSender := tfController.IsSender(uint8(tag))
+	tableID := pktIn.TableId
+
+	if isSender {
+		ob := new(opsv1alpha1.Observation)
+		ob.Component = opsv1alpha1.SpoofGuard
+		ob.Action = opsv1alpha1.Forwarded
+		obs = append(obs, *ob)
+	} else {
+		ob := new(opsv1alpha1.Observation)
+		ob.Component = opsv1alpha1.Forwarding
+		ob.Action = opsv1alpha1.Received
+		ob.ComponentInfo = tableIDToComponentInfo[0]
```

Comment:
If you just need table name, then you can call GetFlowTableName(tableNumber) to get the name. No need to define a separate map. GetFlowTableName now iterate a slice to look up for a table, but since the slice is small I think performance should not be an issue.

Reply:
Thanks. Removed tableIDToComponentInfo[] and moved to GetFlowTableName().